### PR TITLE
feat: update pass constructor to allow partial passes from start/end

### PR DESCRIPTION
**Pull request by vishwa2710**
- [x] Add tests to cover the remaining edge cases

The main lift here is having explicit definitions for all the parameters for a Pass. This was necessary as a Pass can be partial from both sides:
```
|        ----------------|
|--------------          |
```
Therefore we want explicit definitions for the ascending node, north point, descending node, south point and the pass break.
Where pass break is the ascending node for the _next_ pass.

This is also necessary to properly define whether the pass is complete or partial.

Note: While there are some breaking changes here, since this function was recently released and is not used anywhere (yet) we don't necessarily have to do a major version increase here.

### CODE REVIEW

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit.cpp [hunk 1]
```diff
@@ -245,11 +247,12 @@ Pass Orbit::getPassWithRevolutionNumber(const Integer& aRevolutionNumber) const
             Integer currentRevolutionNumber = currentPass.isDefined() ? currentPass.getRevolutionNumber()
                                                                       : this->modelPtr_->getRevolutionNumberAtEpoch();
             Instant previousInstant = currentPass.isDefined()
-                                        ? (currentPass.getInterval().accessEnd() + Duration::Microseconds(1.0))
+                                        ? (currentPass.accessInstantAtPassBreak() + Duration::Microseconds(1.0))
```
Comment (kyle-cochran) [inline src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit.cpp:250]: Sanity check:

From our conversation, I thought I remembered that `passBreak` is the instant immediately after the spacecraft crosses the ascending node. If that's true, would this logic make the next pass start one step after the actual start of the pass?

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit.cpp [hunk 2]
```diff
@@ -1034,13 +1036,12 @@ Map<Index, Pass> Orbit::GeneratePassMap(const Array<State>& aStateArray, const I
...
+    Array<Pair<Index, Pass>> passMap;
 
-    Integer revolutionNumber = anInitialRevolutionNumber;
-    Instant previousPassEndInstant = Instant::Undefined();
-
-    Index currentIndex = 0;
-    State const* previousStatePtr = nullptr;
+    if (aStateArray.getSize() < 2)
+    {
+        return passMap;
```
Comment (kyle-cochran) [inline src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit.cpp:1043]:
```suggestion
        throw ostk::core::error::Runtime("Too few states to compute passes. n_states = {}", aStateArray.getSize());
```

Returning an uninitialized value seems kinda sketch. Maybe good to just error here instead?

Comment (kyle-cochran) [reply src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit.cpp:1043]: Especially since that's the way we handle states not being in chrono order on line 1050 below

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Pass.cpp [hunk 3]
```diff
@@ -15,23 +15,70 @@ namespace orbit
...
+    if (!anInstantAtNorthPoint.isDefined() && !anInstantAtSouthPoint.isDefined() &&
+        !anInstantAtDescendingNode.isDefined() && !anInstantAtAscendingNode.isDefined() &&
+        !anInstantAtPassBreak.isDefined() && !revolutionNumber_.isDefined())
+    {
+        type_ = Pass::Type::Undefined;
+    }
+    else if (!anInstantAtNorthPoint.isDefined() || !anInstantAtSouthPoint.isDefined() ||
+        !anInstantAtDescendingNode.isDefined() || !anInstantAtAscendingNode.isDefined() ||
+        !anInstantAtPassBreak.isDefined())
+    {
```
Comment (kyle-cochran) [inline src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Pass.cpp:41]:
```suggestion
    else if ((!anInstantAtNorthPoint.isDefined() || !anInstantAtSouthPoint.isDefined() ||
        !anInstantAtDescendingNode.isDefined() || !anInstantAtAscendingNode.isDefined() ||
        !anInstantAtPassBreak.isDefined()) && revolutionNumber_.isDefined())
    {
```

Because revolution number always has to be defined right?

Otherwise i think this allows revolution number to be undefined and have a pass with a `Partial` type

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit.cpp (hunk 1 above)
Comment (vishwa2710) [reply src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit.cpp:250]: `passBreak` is the ascending node (within accuracy). Since `passBreak` is at ascending node, We want to start propagating from a microsecond after so that we don't start before the ascending node.

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit.cpp (hunk 2 above)
Comment (vishwa2710) [reply src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit.cpp:1043]: Yeah I went back and forth on this. I think raising an error is fair, users can handle it.

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Pass.cpp (hunk 3 above)
Comment (vishwa2710) [reply src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Pass.cpp:41]: Yeah good catch!

**Review by kyle-cochran — APPROVED** [review]
